### PR TITLE
Ingester prevent writes to large traces even after flushing

### PR DESCRIPTION
**Pull request by mdisibio**
**What this PR does**:
This PR updates the ingester to keep track of traces that exceeded max size and reject further writes to them, even after they are flushed to disk. The list of traces is reset when cutting the head block. 

**Which issue(s) this PR fixes**:
Fixes part of #1133 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -32,6 +32,7 @@
 * [BUGFIX] Ensure that the admin client jsonnet has correct S3 bucket property. (@hedss)
 * [BUGFIX] Publish tenant index age correctly for tenant index writers. [#1146](https://github.com/grafana/tempo/pull/1146) (@joe-elliott)
 * [BUGFIX] Ingester startup panic `slice bounds out of range` [#1195](https://github.com/grafana/tempo/issues/1195) (@mdisibio)
+* [BUGFIX] Prevent writes to large traces even after flushing to disk [#1199](https://github.com/grafana/tempo/pull/1199) (@mdisibio)
```
Comment (joe-elliott) [inline CHANGELOG.md:35]: enhancement? i think this is just an improvement over what already existed and we'll never get it perfect.

File: modules/ingester/instance.go [hunk 2]
```diff
@@ -140,9 +161,6 @@ func (i *instance) Push(ctx context.Context, req *tempopb.PushRequest) error {
 		return status.Errorf(codes.FailedPrecondition, "%s max live traces exceeded for tenant %s: %v", overrides.ErrorPrefixLiveTracesExceeded, i.instanceID, err)
 	}
 
-	i.tracesMtx.Lock()
-	defer i.tracesMtx.Unlock()
-
 	id, err := pushRequestTraceID(req)
```
Comment (joe-elliott) [inline modules/ingester/instance.go:164]: i wonder if we should just delete `Push()`? it hasn't been used for 3 or 4 releases.

File: modules/ingester/trace.go [hunk 3]
```diff
@@ -54,7 +50,8 @@ func (t *trace) Push(_ context.Context, instanceID string, trace []byte, searchD
 	if t.maxBytes != 0 {
 		reqSize := len(trace)
 		if t.currentBytes+reqSize > t.maxBytes {
-			return status.Errorf(codes.FailedPrecondition, "%s max size of trace (%d) exceeded while adding %d bytes to trace %s", overrides.ErrorPrefixTraceTooLarge, t.maxBytes, reqSize, hex.EncodeToString(t.traceID))
+			//return status.Errorf(codes.FailedPrecondition, "%s max size of trace (%d) exceeded while adding %d bytes to trace %s", overrides.ErrorPrefixTraceTooLarge, t.maxBytes, reqSize, hex.EncodeToString(t.traceID))
```
Comment (joe-elliott) [inline modules/ingester/trace.go:53]: remove comment?

**Review by joe-elliott — APPROVED** [review]
minor thoughts. lgtm

File: CHANGELOG.md (hunk 1 above)
Comment (annanay25) [inline CHANGELOG.md:35]: I think another solution that might cover more cases is returning an error from `CombineTraces()` as soon as the pieces sum up to > maxbytes. That will help fix compaction as well as querying.

File: modules/ingester/instance.go (hunk 2 above)
Comment (mdisibio) [reply modules/ingester/instance.go:164]: Agree but didn't seem in-scope for this PR.

File: CHANGELOG.md (hunk 1 above)
Comment (mdisibio) [reply CHANGELOG.md]: Agree, we can explore that in a separate PR.  Let's continue collecting thoughts in #1133 and #976.

**Pull request merged**